File: targon/epistula.py
import json
from hashlib import sha256
from uuid import uuid4
from math import ceil
from typing import Annotated, Any, Dict, List, Optional, Union

import time
from substrateinterface import Keypair


def generate_header(
    hotkey: Keypair,
    body: Union[Dict[Any, Any], List[Any]],
    signed_for: Optional[str] = None,
) -> Dict[str, Any]:
    timestamp = round(time.time() * 1000)
    timestampInterval = ceil(timestamp / 1e4) * 1e4
    uuid = str(uuid4())
    headers = {
        "Epistula-Version": str(2),
        "Epistula-Timestamp": str(timestamp),
        "Epistula-Uuid": uuid,
        "Epistula-Signed-By": hotkey.ss58_address,
        "Epistula-Request-Signature": "0x"
        + hotkey.sign(
            f"{sha256(json.dumps(body).encode('utf-8'))}.{uuid}.{timestamp}.{signed_for}"
        ).hex(),
    }
    if signed_for:
        headers["Epistula-Signed-For"] = signed_for
        headers["Epistula-Secret-Signature-0"] = (
            "0x" + hotkey.sign(str(timestampInterval - 1) + "." + signed_for).hex()
        )
        headers["Epistula-Secret-Signature-1"] = (
            "0x" + hotkey.sign(str(timestampInterval) + "." + signed_for).hex()
        )
        headers["Epistula-Secret-Signature-2"] = (
            "0x" + hotkey.sign(str(timestampInterval + 1) + "." + signed_for).hex()
        )
    return headers


def verify_signature(
    signature, body: bytes, timestamp, uuid, signed_for, sender, now
) -> Optional[Annotated[str, "Error Message"]]:
    if not isinstance(signature, str):
        return "Invalid Signature"
    timestamp = int(timestamp)
    if not isinstance(timestamp, int):
        return "Invalid Timestamp"
    if not isinstance(sender, str):
        return "Invalid Sender key"
    if not isinstance(signed_for, str):
        return "Invalid receiver key"
    if not isinstance(uuid, str):
        return "Invalid uuid"
    if not isinstance(body, bytes):
        return "Body is not of type bytes"
    ALLOWED_DELTA_MS = 5000
    keypair = Keypair(ss58_address=sender)
    if timestamp + ALLOWED_DELTA_MS < now:
        return "Request is too stale"
    verified = keypair.verify(
        f"{sha256(body)}.{uuid}.{timestamp}.{signed_for}", signature
    )
    if not verified:
        return "Signature Mismatch"
    return None

File: validator/epistula.go
package validator

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"math"
	"time"

	"github.com/ChainSafe/go-schnorrkel"
	"github.com/google/uuid"
)

func signMessage(message []byte, public string, private string) (string, error) {
	// Signs a message via schnorrkel pub and private keys
	var pubk [32]byte
	data, err := hex.DecodeString(public)
	if err != nil {
		return "", err
	}
	copy(pubk[:], data)

	var prik [32]byte
	data, err = hex.DecodeString(private)
	if err != nil {
		return "", err
	}
	copy(prik[:], data)

	priv := schnorrkel.SecretKey{}
	priv.Decode(prik)
	pub := schnorrkel.PublicKey{}
	pub.Decode(pubk)
	signingCtx := []byte("substrate")
	signingTranscript := schnorrkel.NewSigningContext(signingCtx, message)
	sig, _ := priv.Sign(signingTranscript)
	sigEncode := sig.Encode()
	out := hex.EncodeToString(sigEncode[:])
	return "0x" + out, nil
}

func sha256Hash(str []byte) string {
	h := sha256.New()
	h.Write(str)
	sum := h.Sum(nil)
	return hex.EncodeToString(sum)
}

// Takes sender ss58, sender public, sender private, receiver ss58 and body
func GetEpistulaHeaders(sSS58, sPub, sPriv, rSS58 string, body []byte) (map[string]string, error) {
	timestamp := time.Now().UnixMilli()
	uuid := uuid.New().String()
	timestampInterval := int64(math.Ceil(float64(timestamp) / 1e4))
	bodyHash := sha256Hash(body)
	message := fmt.Sprintf("%s.%s.%d.%s", bodyHash, uuid, timestamp, rSS58)
	requestSignature, err := signMessage([]byte(message), sPub, sPriv)
	if err != nil {
		return nil, err
	}

	s1, _ := signMessage(fmt.Appendf([]byte{}, "%d.%s", timestampInterval-1, sSS58), sPub, sPriv)
	s2, _ := signMessage(fmt.Appendf([]byte{}, "%d.%s", timestampInterval, sSS58), sPub, sPriv)
	s3, _ := signMessage(fmt.Appendf([]byte{}, "%d.%s", timestampInterval+1, sSS58), sPub, sPriv)

	headers := map[string]string{
		"Epistula-Version":            "2",
		"Epistula-Timestamp":          fmt.Sprintf("%d", timestamp),
		"Epistula-Uuid":               uuid,
		"Epistula-Signed-By":          sSS58,
		"Epistula-Signed-For":         rSS58,
		"Epistula-Request-Signature":  requestSignature,
		"Epistula-Secret-Signature-0": s1,
		"Epistula-Secret-Signature-1": s2,
		"Epistula-Secret-Signature-2": s3,
		"Content-Type":                "application/json",
		"Connection":                  "keep-alive",
	}

	return headers, nil
}
